File: OfflineDemo-Java/app/src/main/java/com/kaltura/player/offlinedemo/MainActivity.java
package com.kaltura.player.offlinedemo;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.os.SystemClock;
import android.text.TextUtils;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import com.google.android.material.snackbar.Snackbar;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.kaltura.playkit.PKDrmParams;
import com.kaltura.playkit.PKLog;
import com.kaltura.playkit.PKMediaEntry;
import com.kaltura.playkit.PKMediaSource;
import com.kaltura.playkit.Utils;
import com.kaltura.tvplayer.KalturaPlayer;
import com.kaltura.tvplayer.OfflineManager;
import com.kaltura.tvplayer.offline.OfflineManagerSettings;

import java.io.IOException;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

public class MainActivity extends AppCompatActivity {

    private PKLog log = PKLog.get("MainActivity");
    private OfflineManager manager;
    private HashMap<String, Item> itemMap = new HashMap<>();
    private Long startTime = 0L;
    private ListView assetListView;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        assetListView = findViewById(R.id.assetList);

        String itemsJson = Utils.readAssetToString(this, "items.json");

        Gson gson = new Gson();

        Type itemJsonType = new TypeToken<ArrayList<ItemJSON>>(){}.getType();

        List<ItemJSON> itemJson = gson.fromJson(itemsJson, itemJsonType);

        List<Item> testItems = new ArrayList<>();

        for (int index = 0; index < itemJson.size(); index++) {
            testItems.add(itemJson.get(index).toItem());
        }

        manager = OfflineManager.getInstance(this);
        manager.setOfflineManagerSettings(new OfflineManagerSettings().setHlsAudioBitrateEstimation(OfflineManagerSettings.DEFAULT_HLS_AUDIO_BITRATE_ESTIMATION));

        manager.setAssetStateListener(new OfflineManager.AssetStateListener() {

            @Override
            public void onStateChanged(@NonNull String assetId, @NonNull OfflineManager.DownloadType downloadType, @NonNull OfflineManager.AssetInfo assetInfo) {
                toast("onStateChanged");
                updateItemStatus(assetId);
            }

            @Override
            public void onAssetRemoved(@NonNull String assetId, @NonNull OfflineManager.DownloadType downloadType) {
                toast("onAssetRemoved");
                updateItemStatus(assetId);
            }

            @Override
            public void onAssetRemoveError(@NonNull String assetId, @NonNull OfflineManager.DownloadType downloadType, @NonNull Exception error) {
                toast("onAssetRemoveError");
                updateItemStatus(assetId);
            }

            @Override
            public void onAssetDownloadFailed(@NonNull String assetId, @NonNull OfflineManager.DownloadType downloadType, @NonNull Exception error) {
                toastLong("Download of" + error + "failed:" + error);
                updateItemStatus(assetId);
            }

            @Override
            public void onAssetDownloadComplete(@NonNull String assetId, @NonNull OfflineManager.DownloadType downloadType) {
                log.d("onAssetDownloadComplete");
                log.d("onAssetDownloadComplete:" + (SystemClock.elapsedRealtimeNanos() - startTime));
                toast("Complete");
                updateItemStatus(assetId);
            }

            @Override
            public void onAssetPrefetchComplete(@NonNull String assetId, @NonNull OfflineManager.DownloadType downloadType) {
                log.d("onAssetPrefetchComplete $assetId totalDownloadTime: ${SystemClock.elapsedRealtimeNanos() - startTime}");

                if (downloadType == OfflineManager.DownloadType.FULL) {
                    toast("Complete");
                } else {
                    toast("Prefetched id =  $assetId");
                }
                updateItemStatus(assetId);
            }

            @Override
            public void onAssetDownloadPending(@NonNull String assetId, @NonNull OfflineManager.DownloadType downloadType) {
                updateItemStatus(assetId);
            }

            @Override
            public void onAssetDownloadPaused(@NonNull String assetId, @NonNull OfflineManager.DownloadType downloadType) {
                toast("Paused");
                updateItemStatus(assetId);
            }

            @Override
            public void onRegistered(@NonNull String assetId, @NonNull OfflineManager.DrmStatus drmStatus) {
                toast("onRegistered:" + drmStatus.currentRemainingTime + "seconds left");
                updateItemStatus(assetId);
            }

            @Override
            public void onRegisterError(@NonNull String assetId, @NonNull OfflineManager.DownloadType downloadType, @NonNull Exception error) {
                toastLong("onRegisterError:" + assetId + " " + error);
                updateItemStatus(assetId);
            }
        });

        manager.setDownloadProgressListener((assetId, bytesDownloaded, totalBytesEstimated, percentDownloaded) -> {
            log.d("[progress] " + assetId +": " + (bytesDownloaded / 1000) + "/" + (totalBytesEstimated / 1000));

            if (itemMap != null && itemMap.get(assetId) != null) {
                Objects.requireNonNull(itemMap.get(assetId)).setBytesDownloaded(bytesDownloaded);
                Objects.requireNonNull(itemMap.get(assetId)).setPercentDownloaded(percentDownloaded);
            } else {
                return;
            }

            runOnUiThread(() ->{
                assetListView.invalidateViews();
            });
        });

        ArrayAdapter<Item> itemArrayAdapter = new ArrayAdapter(this, android.R.layout.simple_list_item_1);

        for (Item item : testItems) {
            if (item != null) {
                itemArrayAdapter.add(item);
                itemMap.put(item.id(), item);
            }
        }

        assetListView.setAdapter(itemArrayAdapter);

        assetListView.setOnItemClickListener((parent, view, position, id) -> {
            showActionsDialog((Item)parent.getItemAtPosition(position), position);
        });

        try {
            manager.start(() -> {
                log.d("manager started");
                for (Map.Entry<String, Item> stringItemEntry : itemMap.entrySet()) {
                    Item itemEntry = stringItemEntry.getValue();
                    itemEntry.setAssetInfo(manager.getAssetInfo(itemEntry.id()));
                }

                runOnUiThread(() -> {
                    assetListView.invalidateViews();
                });
            });
        } catch (IOException e) {
            log.e("IOException in Offline Manager start");
            e.printStackTrace();
        }
    }

    private void updateItemStatus(String assetId) {
        if (itemMap != null && itemMap.get(assetId) != null) {
            updateItemStatus(Objects.requireNonNull(itemMap.get(assetId)));
        }
    }

    private void showActionsDialog(Item item, int position) {
        String[] items = new String[]{"Prepare", "Start", "Pause", "Play-Offline", "Play-Online", "Remove", "Status", "Update"};

        new AlertDialog.Builder(MainActivity.this).setItems(items, (dialog, which) -> {
            switch (which) {
                case 0:
                    doPrepare(item);
                    break;
                case 1:
                    doStart(item);
                    break;
                case 2:
                    doPause(item);
                    break;
                case 3:
                    doOfflinePlayback(item);
                    break;
                case 4:
                    doOnlinePlayback(item, position);
                    break;
                case 5:
                    doRemove(item);
                    break;
                case 6:
                    doStatus(item);
                    break;
                case 7:
                    updateItemStatus(item);
                    break;
            }
        }).show();
    }

    private void updateItemStatus(Item item) {
        item.setAssetInfo(manager.getAssetInfo(item.id()));
        runOnUiThread(() -> assetListView.invalidateViews());
    }

    private void doStatus(Item item) {
        if (item instanceof KalturaItem) {
            OfflineManager.DrmStatus drmStatus = manager.getDrmStatus(item.id());

            if (drmStatus.isClear()) {
                toastLong("Clear");
                return;
            }

            String msg;

            if (drmStatus.isValid()) {
                msg = "Valid; will expire in " + drmStatus.currentRemainingTime + " seconds";
            } else {
                msg = "Expired or Error";
            }

            Snackbar.make(assetListView, msg, Snackbar.LENGTH_LONG).setDuration(5000).setAction("Renew", v -> {
                manager.setKalturaParams(KalturaPlayer.Type.ovp, ((KalturaItem) item).getPartnerId());
                manager.renewDrmAssetLicense(item.id(), ((KalturaItem) item).mediaOptions(), new OfflineManager.MediaEntryCallback() {
                    @Override
                    public void onMediaEntryLoaded(@NonNull String assetId, @NonNull PKMediaEntry mediaEntry) {
                        reduceLicenseDuration(mediaEntry, 300);
                    }

                    @Override
                    public void onMediaEntryLoadError(@NonNull Exception error) {
                        toastLong("onMediaEntryLoadError: " + error);
                    }
                });
            }).show();
        }
    }

    private void doRemove(Item item) {
        if (item.getAssetInfo() != null) {
            manager.removeAsset(item.getAssetInfo().getAssetId());
        } else {
            return;
        }
        updateItemStatus(item);
    }

    private void doOfflinePlayback(Item item) {
        Intent playerActivityIntent = new Intent(this, PlayActivity.class);
        if (item.getAssetInfo() != null) {
            playerActivityIntent.setData(Uri.parse(item.getAssetInfo().getAssetId()));
        }
        startActivity(playerActivityIntent);
    }

    private void doOnlinePlayback(Item item, int position) {
        Intent intent = new Intent(this, PlayActivity.class);
        Bundle bundle = new Bundle();
        bundle.putBoolean("isOnlinePlayback", true);
        bundle.putInt("position", position);
        if (item instanceof OTTItem) {
            bundle.putInt("partnerId", ((OTTItem) item).getPartnerId());
        }

        if (item instanceof OVPItem) {
            bundle.putInt("partnerId", ((OVPItem) item).getPartnerId());
        }

        intent.putExtra("assetBundle", bundle);

        startActivity(intent);
    }

    private void doPause(Item item) {
        if (item != null && item.id() != null) {
            manager.pauseAssetDownload(item.id());
            updateItemStatus(item);
        }
    }

    private void doStart(Item item) {
        log.d("doStart");
        if (item.getAssetInfo() == null) {
            return;
        }

        OfflineManager.AssetInfo assetInfo = item.getAssetInfo();

        startTime = SystemClock.elapsedRealtime();
        manager.startAssetDownload(assetInfo);
        updateItemStatus(item);
    }

    private void doPrepare(Item item) {

        if (item instanceof OTTItem) {
            manager.setKalturaParams(KalturaPlayer.Type.ott, ((OTTItem) item).getPartnerId());
            manager.setKalturaServerUrl(((OTTItem) item).getServerUrl());
        }

        if (item instanceof OVPItem) {
            manager.setKalturaParams(KalturaPlayer.Type.ovp, ((KalturaItem) item).getPartnerId());
            manager.setKalturaServerUrl(((KalturaItem) item).getServerUrl());
        }

        OfflineManager.PrepareCallback prepareCallback = new OfflineManager.PrepareCallback() {
            @Override
            public void onPrepared(@NonNull String assetId, @NonNull OfflineManager.AssetInfo assetInfo, @Nullable Map<OfflineManager.TrackType, List<OfflineManager.Track>> selected) {
                item.setAssetInfo(assetInfo);
                runOnUiThread(() -> {
                    Snackbar.make(assetListView, "Prepared", Snackbar.LENGTH_LONG).setDuration(5000).setAction("Start", v -> doStart(item)).show();
                    assetListView.invalidateViews();
                });
            }

            @Override
            public void onPrepareError(@NonNull String assetId, @NonNull Exception error) {
                toastLong("onPrepareError: " + error);
            }

            @Override
            public void onMediaEntryLoadError(@NonNull Exception error) {
                toastLong("onMediaEntryLoadError: " + error);
            }

            @Override
            public void onMediaEntryLoaded(@NonNull String assetId, @NonNull PKMediaEntry mediaEntry) {
                reduceLicenseDuration(mediaEntry, 300);
            }

            @Override
            public void onSourceSelected(@NonNull String assetId, @NonNull PKMediaSource source, @Nullable PKDrmParams drmParams) {

            }
        };

        OfflineManager.SelectionPrefs defaultPrefs = new OfflineManager.SelectionPrefs();
        defaultPrefs.videoHeight = 300;
        defaultPrefs.videoBitrate = 600000;
        defaultPrefs.videoWidth = 400;
        defaultPrefs.allAudioLanguages = true;
        defaultPrefs.allTextLanguages = true;
        defaultPrefs.allowInefficientCodecs = false;

        if (item instanceof KalturaItem) {
            if (!TextUtils.isEmpty(((KalturaItem) item).getServerUrl())) {
                manager.setKalturaServerUrl(((KalturaItem) item).getServerUrl());
            }
            manager.prepareAsset(((KalturaItem) item).mediaOptions(), item.getSelectionPrefs() != null ? item.getSelectionPrefs() : defaultPrefs, prepareCallback);
        } else {
            if (item.getEntry() != null) {
                manager.prepareAsset(item.getEntry(), item.getSelectionPrefs() != null ? item.getSelectionPrefs() : defaultPrefs, prepareCallback);
            }
        }

    }

    private void reduceLicenseDuration(PKMediaEntry mediaEntry, int seconds) {
        for (PKMediaSource source : mediaEntry.getSources()) {
            if (source.hasDrmParams()) {
                for (PKDrmParams params : source.getDrmData()) {
                    params.setLicenseUri(params.getLicenseUri() + "&rental_duration=" + seconds);
                }
            }
        }
    }

    private void toastLong(String msg) {
        runOnUiThread(() -> Toast.makeText(MainActivity.this, msg, Toast.LENGTH_LONG).show());
    }

    private void toast(String msg) {
        runOnUiThread(() -> Toast.makeText(this, msg, Toast.LENGTH_SHORT).show());
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.action_settings) {
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (manager != null) {
            // Removing listeners by setting it to null
            manager.setAssetStateListener(null);
            manager.setDownloadProgressListener(null);
            manager.stop();
        }
    }
}

File: OfflineDemo/app/src/main/java/com/kaltura/player/offlinedemo/MainActivity.kt
package com.kaltura.player.offlinedemo

import android.annotation.SuppressLint
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.SystemClock
import android.text.TextUtils
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import com.google.android.material.snackbar.Snackbar
import com.google.gson.Gson
import com.kaltura.playkit.*
import com.kaltura.tvplayer.KalturaPlayer
import com.kaltura.tvplayer.MediaOptions
import com.kaltura.tvplayer.OfflineManager
import com.kaltura.tvplayer.offline.OfflineManagerSettings
import kotlinx.android.synthetic.main.activity_main.*
import java.util.*

fun String.fmt(vararg args: Any?): String = java.lang.String.format(Locale.ROOT, this, *args)

@SuppressLint("ParcelCreator")
object NULL : KalturaItem(0, "", null, null) {
    override fun id(): String = TODO()
    override fun mediaOptions(): MediaOptions = TODO()
}

class MainActivity : AppCompatActivity() {

    private lateinit var itemArrayAdapter: ArrayAdapter<Item>
    private lateinit var manager: OfflineManager
    private val itemMap = mutableMapOf<String, Item>()

    var startTime = 0L

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
//        setSupportActionBar(toolbar)

        val itemsJson = Utils.readAssetToString(this, "items.json")

        val gson = Gson()
        val items = gson.fromJson(itemsJson, Array<ItemJSON>::class.java)

        val testItems = items.map { it.toItem() }

        manager = OfflineManager.getInstance(this, OfflineManager.OfflineProvider.DTG)
        manager.setOfflineManagerSettings(OfflineManagerSettings().setHlsAudioBitrateEstimation(64000))

        manager.setAssetStateListener(object : OfflineManager.AssetStateListener {
            override fun onAssetDownloadFailed(assetId: String, downloadType: OfflineManager.DownloadType, error: Exception) {
                toastLong("Download of $error failed: $error")
                updateItemStatus(assetId)
            }

            override fun onAssetDownloadComplete(assetId: String, downloadType: OfflineManager.DownloadType, ) {
                log.d("onAssetDownloadComplete")

                log.d("onAssetDownloadComplete: ${SystemClock.elapsedRealtimeNanos() - startTime}")
                toast("Complete")
                updateItemStatus(assetId)
            }

            override fun onAssetDownloadPending(assetId: String, downloadType: OfflineManager.DownloadType, ) {
                updateItemStatus(assetId)
            }

            override fun onAssetDownloadPaused(assetId: String, downloadType: OfflineManager.DownloadType, ) {
                toast("Paused")
                updateItemStatus(assetId)
            }

            override fun onRegistered(assetId: String, drmStatus: OfflineManager.DrmStatus) {
                toast("onRegistered: ${drmStatus.currentRemainingTime} seconds left")
                updateItemStatus(assetId)
            }

            override fun onRegisterError(assetId: String, downloadType: OfflineManager.DownloadType, error: Exception) {
                toastLong("onRegisterError: $assetId $error")
                updateItemStatus(assetId)
            }

            override fun onUnRegisterError(assetId: String, downloadType: OfflineManager.DownloadType, error: java.lang.Exception) {
                toastLong("onUnRegisterError: $assetId $error")
            }

            override fun onStateChanged(assetId: String, downloadType: OfflineManager.DownloadType, assetInfo: OfflineManager.AssetInfo) {
                toast("onStateChanged")
                updateItemStatus(assetId)
            }

            override fun onAssetRemoved(assetId: String, downloadType: OfflineManager.DownloadType, ) {
                toast("onAssetRemoved")
                updateItemStatus(assetId)
            }

            override fun onAssetRemoveError(
                assetId: String,
                downloadType: OfflineManager.DownloadType,
                error: java.lang.Exception
            ) {
                toast("Error Asset Was Not Removed")
            }

            override fun onAssetPrefetchComplete(
                assetId: String,
                downloadType: OfflineManager.DownloadType
            ) {
            }
        })

        manager.setDownloadProgressListener { assetId, bytesDownloaded, totalBytesEstimated, percentDownloaded ->
            log.d("[progress] $assetId: ${bytesDownloaded / 1000} / ${totalBytesEstimated / 1000}")

            val item = itemMap[assetId] ?: return@setDownloadProgressListener
            item.bytesDownloaded = bytesDownloaded
            item.percentDownloaded = percentDownloaded

            runOnUiThread {
                assetList.invalidateViews()
            }
        }

        itemArrayAdapter = ArrayAdapter(this, android.R.layout.simple_list_item_1)

        testItems.filter { it != NULL }.forEach {
            itemArrayAdapter.add(it)
            itemMap[it.id()] = it
        }

        assetList.adapter = this.itemArrayAdapter

        assetList.setOnItemClickListener { av: AdapterView<*>, _: View, pos: Int, _: Long ->
            showActionsDialog(av.getItemAtPosition(pos) as Item, pos)
        }

        manager.start {
            log.d("manager started")
            itemMap.values.forEach {
                it.assetInfo = manager.getAssetInfo(it.id())
            }

            runOnUiThread(assetList::invalidateViews)
        }
    }

    private fun updateItemStatus(assetId: String) {
        updateItemStatus(itemMap[assetId] ?: return)
    }

    private fun showActionsDialog(item: Item, position: Int) {
        val items = arrayOf("Prepare", "Start", "Pause", "Play-Offline", "Play-Online", "Remove", "Status", "Update")
        AlertDialog.Builder(this).setItems(items) { _, i ->
            when (i) {
                0 -> doPrepare(item)
                1 -> doStart(item)
                2 -> doPause(item)
                3 -> doOfflinePlayback(item)
                4 -> doOnlinePlayback(item, position)
                5 -> doRemove(item)
                6 -> doStatus(item)
                7 -> updateItemStatus(item)
            }
        }.show()
    }

    private fun updateItemStatus(item: Item) {
        item.assetInfo = manager.getAssetInfo(item.id())
        runOnUiThread { assetList.invalidateViews() }
    }

    private fun doStatus(item: Item) {

        if (item !is KalturaItem) {
            toastLong("Not applicable")
            return
        }

        val drmStatus = manager.getDrmStatus(item.id())

        if (drmStatus.isClear) {
            toastLong("Clear")
            return
        }

        val msg = when {
            drmStatus.isValid -> "Valid; will expire in " + drmStatus.currentRemainingTime + " seconds"
            else -> "Expired or Error"
        }

        snackbar(msg, "Renew") {
            manager.setKalturaParams(KalturaPlayer.Type.ovp, item.partnerId)
            manager.renewDrmAssetLicense(item.id(), item.mediaOptions(), object: OfflineManager.MediaEntryCallback {
                override fun onMediaEntryLoaded(assetId: String, downloadType: OfflineManager.DownloadType, mediaEntry: PKMediaEntry) {
                   // reduceLicenseDuration(mediaEntry, 300)
                }

                override fun onMediaEntryLoadError(downloadType: OfflineManager.DownloadType, error: Exception) {
                    toastLong("onMediaEntryLoadError: $error")
                }
            })
        }
    }

    private fun doRemove(item: Item) {
        manager.removeAsset(item.assetInfo?.assetId ?: return)
        updateItemStatus(item)
    }

    private fun doOfflinePlayback(item: Item) {
        startActivity(Intent(this, PlayActivity::class.java).apply {
            data = Uri.parse(item.assetInfo?.assetId ?: return)
        })
    }

    private fun doOnlinePlayback(item: Item, position: Int) {
        val intent = Intent(this, PlayActivity::class.java)

        val bundle = Bundle()
        bundle.putBoolean("isOnlinePlayback", true)
        bundle.putInt("position", position)

        if (item is OTTItem) {
            bundle.putInt("partnerId", item.partnerId)
        }

        if (item is OVPItem) {
            bundle.putInt("partnerId", item.partnerId)
        }

        intent.putExtra("assetBundle", bundle)

        startActivity(intent)
    }

    private fun doPause(item: Item?) {
        item?.let { it ->
            it.id()?.let { itemId ->
                manager.pauseAssetDownload(itemId)
                updateItemStatus(it)
            }
        }
    }

    private fun doStart(item: Item) {
        log.d("doStart")
        val assetInfo = item.assetInfo ?: return

        startTime = SystemClock.elapsedRealtime()
        manager.startAssetDownload(assetInfo)
        updateItemStatus(item)
    }

    private fun doPrepare(item: Item) {
        if (item is OTTItem) {
            manager.setKalturaParams(KalturaPlayer.Type.ott, item.partnerId)
            manager.setKalturaServerUrl(item.serverUrl)
        }

        if (item is OVPItem){
            manager.setKalturaParams(KalturaPlayer.Type.ovp, item.partnerId)
            manager.setKalturaServerUrl(item.serverUrl)
        }

        val prepareCallback = object : OfflineManager.PrepareCallback {
            override fun onPrepared(
                assetId: String,
                assetInfo: OfflineManager.AssetInfo,
                selected: MutableMap<OfflineManager.TrackType, MutableList<OfflineManager.Track>>?
            ) {
                item.assetInfo = assetInfo
                runOnUiThread {
                    snackbar("Prepared", "Start") {
                        doStart(item)
                    }
                    assetList.invalidateViews()
                }
            }

            override fun onPrepareError(assetId: String, downloadType: OfflineManager.DownloadType, error: Exception) {
                toastLong("onPrepareError: $error")
            }

            override fun onMediaEntryLoadError(downloadType: OfflineManager.DownloadType, error: Exception) {
                toastLong("onMediaEntryLoadError: $error")
            }

            override fun onMediaEntryLoaded(assetId: String, downloadType: OfflineManager.DownloadType, mediaEntry: PKMediaEntry) {
                //reduceLicenseDuration(mediaEntry, 300)
            }

            override fun onSourceSelected(
                assetId: String,
                source: PKMediaSource,
                drmParams: PKDrmParams?
            ) {

            }
        }

        val defaultPrefs = OfflineManager.SelectionPrefs().apply {
            videoHeight = 300
            videoBitrate = 600000
            videoWidth = 400
            allAudioLanguages = true
            allTextLanguages = true
            allowInefficientCodecs = false
        }

        if (item is KalturaItem) {
            if (!TextUtils.isEmpty(item.serverUrl)) {
                manager.setKalturaServerUrl(item.serverUrl);
            }
            manager.prepareAsset(item.mediaOptions(), item.selectionPrefs ?: defaultPrefs, prepareCallback)
        } else {
            item.entry?.let {
                    entry -> manager.prepareAsset(entry, item.selectionPrefs ?: defaultPrefs, prepareCallback)
            }
        }
    }

    @Suppress("SameParameterValue")
    private fun reduceLicenseDuration(mediaEntry: PKMediaEntry, seconds: Int) {
        for (source in mediaEntry.sources) {
            if (source.hasDrmParams()) {
                for (params in source.drmData) {
                    params.licenseUri = params.licenseUri + "&rental_duration=" + seconds
                }
            }
        }
    }

    private fun snackbar(msg: String, next: String, nextAction: () -> Unit) = runOnUiThread {
        Snackbar.make(assetList, msg, Snackbar.LENGTH_LONG).apply {
            duration = 5000
            setAction(next) {
                nextAction()
            }
            show()
        }
    }

    private fun toast(msg: String) = runOnUiThread { Toast.makeText(this, msg, Toast.LENGTH_SHORT).show() }

    private fun toastLong(msg: String) = runOnUiThread { Toast.makeText(this, msg, Toast.LENGTH_LONG).show() }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        return when (item.itemId) {
            R.id.action_settings -> true
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        manager.let {
            // Removing listeners by setting it to null
            manager.setAssetStateListener(null)
            manager.setDownloadProgressListener(null)
            manager.stop()
        }
    }

    private companion object {
        val log = PKLog.get("MainActivity")
    }
}
